File: sections/72_master_flow_convex_mixed/problems/e_exact_red_forest/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

struct DSU {
    vector<int> parent, size;
    explicit DSU(int n) : parent(n), size(n, 1) {
        iota(parent.begin(), parent.end(), 0);
    }
    int find(int value) {
        while (value != parent[value]) value = parent[value];
        return value;
    }
    bool unite(int a, int b) {
        a = find(a);
        b = find(b);
        if (a == b) return false;
        if (size[a] < size[b]) swap(a, b);
        parent[b] = a;
        size[a] += size[b];
        return true;
    }
};

struct Edge { int u, v, red; ll weight; };

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m, components, wanted;
    cin >> n >> m >> components >> wanted;
    int edge_count = n - components;
    vector<Edge> edges(m);
    for (Edge& edge : edges) {
        char color;
        cin >> edge.u >> edge.v >> edge.weight >> color;
        --edge.u;
        --edge.v;
        edge.red = color == 'R';
    }
    if (edge_count == 0) {
        cout << (wanted == 0 ? "0\n" : "IMPOSSIBLE\n");
        return 0;
    }

    auto run = [&](ll penalty, bool prefer_red) {
        vector<int> order(m);
        iota(order.begin(), order.end(), 0);
        sort(order.begin(), order.end(), [&](int i, int j) {
            ll first = edges[i].weight + penalty * edges[i].red;
            ll second = edges[j].weight + penalty * edges[j].red;
            if (first != second) return first < second;
            return prefer_red ? edges[i].red > edges[j].red
                              : edges[i].red < edges[j].red;
        });
        DSU dsu(n);
        int used = 0;
        int red = 0;
        ll value = 0;
        for (int index : order) {
            const Edge& edge = edges[index];
            if (!dsu.unite(edge.u, edge.v)) continue;
            ++used;
            red += edge.red;
            value += edge.weight + penalty * edge.red;
            if (used == edge_count) break;
        }
        return tuple<int, int, ll>{used, red, value};
    };

    const ll BOUND = 400'000'000'000'000LL;
    auto [minimum_used, minimum_red, ignored1] = run(BOUND, false);
    auto [maximum_used, maximum_red, ignored2] = run(-BOUND, true);
    if (minimum_used != edge_count || maximum_used != edge_count
        || wanted < minimum_red || wanted > maximum_red) {
        cout << "IMPOSSIBLE\n";
        return 0;
    }

    ll low = -BOUND;
    ll high = BOUND;
    while (low < high) {
        ll middle = low + (high - low + 1) / 2;
        auto [used, red, value] = run(middle, true);
        if (red >= wanted)
            low = middle;
        else
            high = middle - 1;
    }
    auto [used, red, modified] = run(low, true);
    cout << modified - low * wanted << '\n';
}

File: sections/72_master_flow_convex_mixed/problems/e_exact_red_forest/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    iterator = iter(data)
    n = int(next(iterator))
    m = int(next(iterator))
    components = int(next(iterator))
    wanted = int(next(iterator))
    edges = []
    for _ in range(m):
        u = int(next(iterator)) - 1
        v = int(next(iterator)) - 1
        weight = int(next(iterator))
        red = next(iterator) == b'R'
        edges.append((u, v, weight, red))
    edge_count = n - components
    if edge_count == 0:
        print(0 if wanted == 0 else "IMPOSSIBLE")
        return

    def run(penalty, prefer_red):
        order = sorted(
            edges,
            key=lambda edge: (
                edge[2] + penalty * edge[3],
                -edge[3] if prefer_red else edge[3],
            ),
        )
        parent = list(range(n))
        size = [1] * n

        def find(value):
            while value != parent[value]:
                value = parent[value]
            return value

        used = red_count = value = 0
        for u, v, weight, red in order:
            u, v = find(u), find(v)
            if u == v:
                continue
            if size[u] < size[v]:
                u, v = v, u
            parent[v] = u
            size[u] += size[v]
            used += 1
            red_count += red
            value += weight + penalty * red
            if used == edge_count:
                break
        return used, red_count, value

    bound = 400_000_000_000_000
    minimum_used, minimum_red, _ = run(bound, False)
    maximum_used, maximum_red, _ = run(-bound, True)
    if (minimum_used != edge_count or maximum_used != edge_count
            or not minimum_red <= wanted <= maximum_red):
        print("IMPOSSIBLE")
        return

    low, high = -bound, bound
    while low < high:
        middle = (low + high + 1) // 2
        if run(middle, True)[1] >= wanted:
            low = middle
        else:
            high = middle - 1
    _, _, modified = run(low, True)
    print(modified - low * wanted)


if __name__ == '__main__':
    main()
